### PR TITLE
Reference tracking

### PR DESCRIPTION
**Pull request by Rohesie**
* Implements proper tools for the extool reference tracking feature.

### CODE REVIEW

**Comment by Azarak** [comment]
ITS HERE!

**Comment by Fox-McCloud** [comment]
The old methodology shouldn't be removed until  the ref checker is properly fixed; at the moment it can return non-sensical results and a lot of false positives---until that issues is addressed, this doesn't seem like the wisest thing to implement.

While the old method is horrifyingly slow, it's also accurate.

**Comment by optimumtact** [comment]
Yes, I don't think we would remove the old method completely ever either, I'm not ready to rely on exttools completely yet.

**Comment by Rohesie** [comment]
The old method is not accurate at all, and is basically useless.
By the time it finishes sweeping the world for references, they have been long gone, hard deleted or just changed state.
Proof of that is that we are full of uncleared references and there's no sane way to check them without extools.

I've tried clearing references before without extools. Just checking what had been hard deleted on the list, feature that is preserved, was the only thing useful. Freezing the game for minutes to find nothing didn't help.

**Comment by optimumtact** [comment]
@MrStonedOne

**Comment by MrStonedOne** [comment]
It seems there are two stories as to which is more accurate.

I will say that you can fix refs going away mid search by setting loop_checks to false at compile time, and you could do more work to follow list references inline to the ref searching rather then as a seperate step to provide info as to where the lists are themselves referenced.

**Comment by Rohesie** [comment]
It will still lock the game for minutes for every single reference search even if all obstacles are overcome.
Few people bother to clean references because it's not a terribly noticeable thing as it is. Running the game to wait the hard delete time plus the search time for every single one, even if you got perfect results, is something I don't see anyone using/doing.

The benefit of the extools system is that it does this instantly. And even then I doubt you'll see too many people hunting those references down to fix them.

If you really insist I can bring back the old system, but it will remain as unused bloat because the byond resources it relies on are simply too inefficient, and the gain from using it is not visible enough.

**Comment by optimumtact** [comment]
they have a point

**Comment by Fox-McCloud** [comment]
Very very few people take seriously the task of de-referencing, that I will absolutely agree.

That said, those who do know their way around things typically--the 2 minute wait is easily avoidable by var editing the GC to fail after something low, like 5-10 seconds.

In any event, I've had generally "ok" luck with utilizing the slow, wonky reference checker without it losing track of refs, though, yes, it takes a bit to print the results and not knowing what list it's referencing is a pain.

That said, the current extools reference tracker is broken really badly; it will return things that absolutely do not have a forward or backwards pointer to an object, and it also suffers from lists not being particularly easy to suss out either (the primary advantage is you can examine the lists).

Until that's fixed, I don't think it's wise to wholesale  switch over to this new tool; it'll generate a lot of false positives and send people down rabbit trails that are literal madness and nonsense.

Few people engage in de-ref hunting, and fewer still are super familiar with the tools, but until extools properly can track deletion of datums (and lists), then forcing everyone to use a faulty, broken tool is going to cause frustration and confusion.

**Comment by Rohesie** [comment]
Legacy system brought back, change requests made.

File: tools/UnstandardnessTestForDM/UnstandardnessTestForDM/obj/x86/Debug/.NETFramework,Version=v4.0,Profile=Client.AssemblyAttributes.cs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+// <autogenerated />
+using System;
```
Comment (AnturK) [inline tools/UnstandardnessTestForDM/UnstandardnessTestForDM/obj/x86/Debug/.NETFramework,Version=v4.0,Profile=Client.AssemblyAttributes.cs:2]: ???

Comment (Rohesie) [reply tools/UnstandardnessTestForDM/UnstandardnessTestForDM/obj/x86/Debug/.NETFramework,Version=v4.0,Profile=Client.AssemblyAttributes.cs]: fixed

**Pull request merged**